Gracefully handle client disconnections

asvoboda opened the pull request:
Currently the server process will log a 500
when clients disconnect. Instead, we can use NGINX style
response codes and log 499.

asvoboda committed:
Gracefully handle client disconnections

Currently the server process will log a 500
when clients disconnect. Instead, we can use NGINX style
response codes and log 499.

bluekeyes commented on baseapp/error.go:
```diff
@@ -45,7 +45,16 @@ func HandleRouteError(w http.ResponseWriter, r *http.Request, err error) {
+	// Either the deadline has passed or the request was canceled
+	// 499 is an NGINX style response code for 'Client Closed Connection'
+	// and is a non-standard, but widely used, HTTP status code
+	if cerr := r.Context().Err(); cerr != nil {
```
After thinking about this a bit, I think we should compare `cerr` to `context.Canceled` and then only return the 499 in that case. It would also be good to remove the error logging in this case (or perhaps log a special message at a lower level).

bluekeyes reviewed (commented):
One small comment about the implementation, but I'd also like to let the version we implemented internally run for a bit more before merging this to make sure there is no unexpected behavior.

akarnani committed:
Cleanup logging logic when client conn drops

If a client disconnects causing an error we should log at a lower level
than error
